third_party/move/move-compiler-v2/src/pipeline/variable_coalescing.rs

```diff
@@ -0,0 +1,235 @@
+    fn new(offset: CodeOffset) -> Self {
+        Self(offset, offset)
+    }
+
+    /// Include the given offset in the live interval, expanding the interval as necessary.
+    fn include(&mut self, offset: CodeOffset) {
+        self.0 = std::cmp::min(self.0, offset);
+        self.1 = std::cmp::max(self.1, offset);
+    }
+}
+
+/// Live interval event of a local, used for sorting.
+enum LiveIntervalEvent {
+    Begin(
+        /* which local? */ TempIndex,
+        /* live interval begins */ CodeOffset,
+        /* live interval length */ usize, // used for tie-breaking
+    ),
+    End(
+        /* which local? */ TempIndex,
+        /* live interval ends */ CodeOffset,
+    ),
+}
+
+impl LiveIntervalEvent {
+    /// Get the code offset at which the event occurs.
+    fn offset(&self) -> CodeOffset {
+        match self {
+            LiveIntervalEvent::Begin(_, offset, _) => *offset,
+            LiveIntervalEvent::End(_, offset) => *offset,
+        }
+    }
+}
+
+pub struct VariableCoalescing {}
+
+impl VariableCoalescing {
+    /// Compute the live intervals of locals in the given function target.
+    /// The result is a vector of live intervals, where the index of the vector is the local.
+    /// If a local has `None` as its live interval, we can ignore the local for the coalescing
+    /// transformation (eg., because it is borrowed or because it is never used): it implies
+    /// that it is the trivial live interval.
+    fn live_intervals(target: &FunctionTarget) -> Vec<Option<LiveInterval>> {
+        let live_var_annotation = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live var annotation is a prerequisite");
+        // Note: we currently exclude all the variables that are borrowed from participating in this
+        // transformation, which is safe. However, we could be more precise in this regard.
+        let borrowed_locals = target.get_borrowed_locals();
+        // Initially, all locals have trivial live intervals.
+        let mut live_intervals = std::iter::repeat_with(|| None)
+            .take(target.get_local_count())
+            .collect::<Vec<_>>();
+        for (offset, live_var_info) in live_var_annotation.0.iter() {
+            live_var_info
+                .after
+                .keys()
+                .chain(live_var_info.before.keys())
+                .filter(|local| !borrowed_locals.contains(local))
+                .for_each(|local| {
+                    // non-borrowed local that is live before and/or after the code offset.
+                    let interval =
+                        live_intervals[*local].get_or_insert_with(|| LiveInterval::new(*offset));
+                    interval.include(*offset);
+                });
+        }
+        live_intervals
+    }
+
+    /// Compute the sorted live interval events of locals in the given function target.
+    /// See implementation comments for the sorting order.
```

Comment:
Why not just add an impl of `Ord` to `LiveInterval`?

Reply:
I had considered it, but:
- I am not sure it buys much as this is the only current use
- this use is special, for eg, if we wanted a `BTreeSet` of `LiveInterval`, we would want to use the normal and simpler default derivable ordering behavior, not this more complex one: so I don't think this is the natural universal order I want to impose.

Reply:
Ok, but it still would be more readable to pull the ordering lambda out and give it a name.

Reply:
This whole PR could benefit from more abstraction which might eliminate the need for so many detailed comments.  It seems correct, but unnecessarily wordy.